.github/workflows/check-release.yml

```diff
@@ -0,0 +1,80 @@
+name: Check Versions and Build-Publish
+
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  check-version-and-build-publish:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Set up Python
+      uses: actions/setup-python@v3
+      with:
+        python-version: 3.8
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install requests build
+
+    - name: Check versions and Build-Publish to PyPI
+      run: |
+        PACKAGE_PATHS=(
+          "kedro-datasets/kedro_datasets"
+          "kedro-telemetry/kedro_telemetry"
+          "kedro-airflow/kedro_airflow"
+          "kedro-docker/kedro_docker"
+        )
+
+        new_release=false
+        for package_path in "${PACKAGE_PATHS[@]}"; do
+          package_name="${package_path%%/*}"
+          package_version=$(python -c "import re; print(re.search(r'__version__\s*=\s*\"([^\"]+)\"', open('$package_path/__init__.py').read()).group(1))")
+          pypi_endpoint="https://pypi.org/pypi/$package_name/$package_version/json"
+          status_code=$(curl --write-out %{http_code} --silent --output /dev/null "$pypi_endpoint")
+          if [ "$status_code" -eq 404 ]; then
+            echo "Starting the release of $package_name $package_version"
+            cd $package_path/..
+            python -m build
+            dist_path=$(pwd)/dist
+            new_release=true
+            break
+          else
+            echo "Skipped: $package_name $package_version already exists on PyPI"
+          fi
+        done
+        echo "new_release=$new_release" >> "$GITHUB_ENV"
+        echo "dist_path=$dist_path" >> "$GITHUB_ENV"
+        echo "package_name=$package_name" >> "$GITHUB_ENV"
+        echo "package_version=$package_version" >> "$GITHUB_ENV"
```

Comment:
You can use Python for scripting without writing Bash.

https://github.com/noklam/github_action/blob/main/.github/workflows/ci.yaml

Reply:
Ideally, I don't expect the release logic only exist in the CI file, it should be a simple call. We should be able to run it locally similarly with `make xxxx`.

Reply:
> One more comment: I think we need to run the test before we publish.
> 
> If all plugins are going to be released. Will they all be published as 1 workflow or 4 separate one? I think it's better to have separate jobs because each of them will have their own testing requirements.
> 
> 1. Check version - see if need to release
> 2. Each plugin gets a separate job for the release (some could fail, some could pass)

This is something I was wondering about, currently, if we were to release several plugins in one PR the script won't work properly as it's designed to only accept 1 plugin release. But would there be a scenario where we would have multiple releases in one PR?

Reply:
> Ideally, I don't expect the release logic only exist in the CI file, it should be a simple call. We should be able to run it locally similarly with `make xxxx`.

I am open to this idea, just to clarify maybe we keep the `Check versions and Build-Publish` logic in a python file similar to `tools/circleci/circleci_release.py`?

Reply:
I think it's not likely we release everything in a PR. But it's good to avoid coupling tightly, we can have a release job that accept plugin as an argument. Then the job will be simply looping through the plugin and decide if it needs to trigger a release job. The same plugin argument  will also be passed to the check plugin job to trigger tests.

Reply:
One way I think this can be done is to have a workflow with 4 jobs each for different plugin that checks if the release should be done eg - `main-updated.yml`
```yaml
name: Main updated

on:
  push:
    branches:
      - main

jobs:
  maybe-trigger-release-docker: 
    steps:
       - name: check if docker needs to be released (__init__.py file updated & new version doesn't exist on PyPi)
       - name: if previous step true, call reusable release workflow
         uses: .github/workflows/release.yml
         with: 
           plugin: docker

  maybe-trigger-release-datasets: 
    steps:
       - name: check if datasets needs to be released (__init__.py file updated & new version doesn't exist on PyPi)
       - name: if previous step true, call reusable release workflow
         uses: .github/workflows/release.yml
         with: 
           plugin: datasets

(same for airflow and telemetry)
``` 
And then have a reusable workflow `release.yml` which accepts plugin name and does the following jobs -
```yaml
on:
  workflow_call:
    inputs:
      plugin:
        type: string

jobs:
  build: 
  test: (call check-plugin.yml with plugin)
  release: 
    needs: [build, test]
```